addons/analytic/models/analytic_distribution_model.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+import json
+
+class AccountAnalyticDistributionModel(models.Model):
+    _name = 'account.analytic.distribution.model'
+    _inherit = 'analytic.mixin'
+    _description = 'Analytic Distribution Model'
+    _rec_name = 'create_date'
+    _order = 'id desc'
+
+    partner_id = fields.Many2one(
+        'res.partner',
+        string='Partner',
+        ondelete='cascade',
+        help="Select a partner for which the analytic distribution will be used (e.g. create new customer invoice or Sales order if we select this partner, it will automatically take this as an analytic account)",
+    )
+    company_id = fields.Many2one(
+        'res.company',
+        string='Company',
+        ondelete='cascade',
+        help="Select a company for which the analytic distribution will be used (e.g. create new customer invoice or Sales order if we select this company, it will automatically take this as an analytic account)",
+    )
+
+    @api.model
+    def _get_distribution(self, vals):
+        """ Returns the distribution model that has the most fields that corresponds to the vals given
+            This method should be called to prefill analytic distribution field on several models """
+        domain = [(fname, 'in', [value, False]) for fname, value in vals.items() if value]
```

Comment:
Why are you checking for False too?
If I have
* If product P, then set Account A1
* If product P and Partner C, then set Account A2

When I set product P on a partner that is not C, I want A1, not A2. With your solution, it might be A2 depending on the order of the records because both are in the domain and have a score of 1.

Reply:
Are you sure? Am I mistaken if I say that if we have the models you mentioned, 
if I have product P, Partner D then search all distribution models that have product = product P or false, partner = partner D or false.
=> Account A1 will be returned while Account A2 shouldn't because wrong partner